src/clj/rems/db/entitlements.clj

```diff
@@ -63,52 +64,62 @@
+                                                    (contains? members userid)
+                                                    (accepted-licenses? application userid))
+                                         resource-id application-resources
+                                         :when (not (contains? resource-ids resource-id))]
+                                     {userid #{resource-id}})
+                                   (apply merge-with union))
+          entitlements-to-remove (->> (for [[userid resource-ids] user-entitlements
+                                            resource-id resource-ids
+                                            :when (or (= :application.state/closed (:application/state application))
+                                                      (not (contains? members userid))
+                                                      (not (accepted-licenses? application userid))
+                                                      (not (contains? application-resources resource-id)))]
+                                        {userid #{resource-id}})
+                                      (apply merge-with union))
+          members-to-update (set
+                             (concat (keys entitlements-to-add)
+                                     (keys entitlements-to-remove)))]
       (when (seq members-to-update)
-        (doseq [user-id members-to-update]
-          (log/info "granting entitlements on application" app-id "to" user-id)
-          (doseq [[resource-id ext-id] (->> (:application/resources application)
-                                            (map (juxt :resource/id :resource/ext-id)))]
+        (log/info "updating entitlements on application" app-id)
+        (doseq [[userid resource-ids] entitlements-to-add]
+          (log/info "granting entitlements on application" app-id "to" userid "resources" resource-ids)
+          (doseq [resource-id (sort resource-ids)]
             (db/add-entitlement! {:application app-id
-                                  :user user-id
+                                  :user userid
+                                  :resource resource-id})
+            (post-entitlements :add (db/get-entitlements {:application app-id :user userid :resource resource-id}))))
+        (doseq [[userid resource-ids] entitlements-to-remove]
+          (log/info "revoking entitlements on application" app-id "to" userid "resources" resource-ids)
+          (doseq [resource-id (sort resource-ids)]
+            (db/end-entitlement! {:application app-id
+                                  :user userid
                                   :resource resource-id})
-            (post-entitlements :add (db/get-entitlements {:application app-id :user user-id :resource ext-id})))))
+            (post-entitlements :remove (db/get-entitlements {:application app-id :user userid :resource resource-id})))))
       members-to-update)))
```

Comment:
is the return value still used?

Reply:
No but I think it's now slightly better when it's explicitly set at something instead of being random return value